### PR TITLE
#410 Access control for interview time table

### LINKED ISSUE #410: Access control for interview time table

### PR DESCRIPTION
**Pull request by n-gb**
Fixed #410 
screenshot from 2017-07-21 16-00-34

### CODE REVIEW

File: components/courses/app/policies/courses/interview/list_policy.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      private
+
+      attr_reader :student, :season
+
+      def student_belongs_to_the_season
+        student.season_id == season.id
+      end
+
+      def student_has_done_test_task
+        student.status == "test_task_done"
```
Comment (sveredyuk) [inline components/courses/app/policies/courses/interview/list_policy.rb]: `student. test_task_done?`

Comment (sveredyuk) [reply components/courses/app/policies/courses/interview/list_policy.rb]: and you already don't need `student_has_done_test_taskz method

File: components/courses/app/models/courses/test_task.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 module Courses
   class TestTask < ApplicationRecord
     self.table_name = 'courses_test_tasks'
-    belongs_to :student, optional: true
+    belongs_to :student
```
Comment (nadiyaka) [inline components/courses/app/models/courses/test_task.rb:4]: How about a situation when a mentor creates a new interview slot, which doesn't belong to a student yet.

File: components/courses/spec/policies/courses/interviews/list_policy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  describe '#allowed?' do
+    context 'student belongs to this season and has approved test task' do
+      it 'allows to pass policy' do
+        expect(policy).to be_allowed
+      end
+    end
+
+    context 'student does not belong to this season' do
+      it 'forbids to pass policy' do
+        wrong_id = 3
```
Comment (sveredyuk) [inline components/courses/spec/policies/courses/interviews/list_policy_spec.rb]: 👎

File: components/courses/app/models/courses/test_task.rb (hunk 2 above)
Comment (sveredyuk) [reply components/courses/app/models/courses/test_task.rb:4]: You mean interview but this is TestTask`

Comment (nadiyaka) [reply components/courses/app/models/courses/test_task.rb:4]: Sorry :(

Comment (NadiaGubernator) [inline components/courses/app/models/courses/test_task.rb]: This isn't interview slot, it's test task. I can't even imagine how there can be test task done without a student.

File: components/courses/app/controllers/courses/interviews_controller.rb [hunk 4]
```diff
@@ -29,5 +30,12 @@ def interviews
     def interview_params
       params.require(:interview).permit(:student_id)
     end
+
+    def execute_list_policy
+      allowed = Interview::ListPolicy.new(current_student, current_season)
+        .allowed?
```
Comment (sveredyuk) [inline components/courses/app/controllers/courses/interviews_controller.rb]:
```ruby
allowed = Interview::ListPolicy.new(
  current_student, current_season
).allowed?
```

File: components/courses/app/controllers/courses/interviews_controller.rb [hunk 5]
```diff
@@ -29,5 +30,12 @@ def interviews
...
+      redirect_to courses_season_path(current_season),
```
Comment (sveredyuk) [inline components/courses/app/controllers/courses/interviews_controller.rb]: `return if allowed`

**Review by sveredyuk — APPROVED** [review]

**Pull request merged**